### ISSUE TITLE
onWillPop callback not called when back button is pressed on WillPopScope Widget

### ISSUE THREAD

**Issue by M1K8** [SchabanBo/qlevar_router#29]
I am trying to override the back button behaviour for use with a `TabController`, where the back button will navigate to a certain tab, yet I am unable to do this using `WillPopScope `& `onWillPop()`. 

I have tried this without qlevar and it would work, yet when using the router `onWillPop()` is never called. Is this intended behaviour, and if so how would I implement this functionality while using qlevar?

The Widget I am try to use is as follows: 

```
Scaffold(
        body: TabBarView(
        controller: tabController,
        children: [
          WillPopScope(child: HomeScreen(), onWillPop: () async {
            tabController.animateTo(0, duration: Duration(milliseconds: 500),);
            return false;
          }),
          WillPopScope(child: FriendScreen(), onWillPop: () async {
            tabController.animateTo(0, duration: Duration(milliseconds: 500),);
            return false;
          }),
          GroupScreen()
        ],),
        bottomNavigationBar: ConvexAppBar(
          controller: tabController,
          style: TabStyle.textIn,
          items: [
            TabItem(icon: Icons.home, title: 'Home'),
            TabItem(icon: Icons.map, title: 'Friends'),
            TabItem(icon: Icons.add, title: 'Groups'),
          ],
          onTap: (int i) => {
            print(QR.history.debug())
          },
        )
    )
```

**Comment by michael-woolsey** [comment]
We're having a similar issue where we're also using the `WillPopScope` widget and it doesn't appear to be getting fired for us either.

Did anyone find a way to get this to work?

**Comment by SchabanBo** [comment]
@michael-woolsey @M1K8 
Did you try `QMiddleware.CanPop`?
See this [example](https://github.com/SchabanBo/qr_samples/blob/main/lib/examples/can_pop.dart)

**Comment by michael-woolsey** [comment]
I will try that example today and let you know! Thanks!

**Comment by michael-woolsey** [comment]
Hello @SchabanBo 

Unfortunately this doesn't appear to solve my problem, while the property can indeed stop the navigation it doesn't help me because I am trying to do a dialog popup to do a confirmation, which requires a Future.

These are the two issues we are having with qlevar, one I have solved the other I'm not sure how to solve yet:

1. When popping scope by pressing the back button on an Android phone, I need a dialog to appear to confirm the navigation.  However, it appears that the `QMiddleware`'s `bool canPop() => true;` property does not accept Futures which is needed for a dialog popup.
2. With a dialog popup, if you press the back button on an Android device the app will navigate instead of simply closing the popup dialog.  I have gotten around this issue using the https://pub.dev/packages/back_button_interceptor package, so this is a non issue for me, however it would be nice if I didn't have to do a work around.

Do you have any idea of a way I can use a Future for my popping of scope?

Thank you,
Michael

**Comment by SchabanBo** [comment]
Hallo @michael-woolsey ,

I will start working on the feature that CanPop will accept future.

**Comment by SchabanBo** [comment]
@michael-woolsey  Is [this](https://qlevar-router.netlify.app/#/parent/child-6) the wanted behavior?
(If the page was not found please clear the browser cash)

**Comment by michael-woolsey** [comment]
It's hard for me to tell in the browser, I was pressing the Chrome browser back button and it was not causing said popup, however I do not know if that's a Chrome thing or not.

The other back arrow button in the web page did indeed popup the dialog as I would like.

**Comment by michael-woolsey** [comment]
I was able to get Chrome to use the back and forth <- and -> keys and the dialog was appearing and preventing navigation if I pressed 'no' ... it appears to be doing showing the dialog as requested. :)

**Comment by SchabanBo** [comment]
@michael-woolsey so with 1.5.0 the the (canPop) is async now. And only the dialog will closed with the back button.
I will wait for your feedback to close this issue 😊.

**Comment by michael-woolsey** [comment]
@SchabanBo - I'm looking into the changes now, it will take me a little bit, but I will test it!

**Comment by michael-woolsey** [comment]
I haven't dug through your code, but if possible a `FutureOr` might be more flexible for folks with your `canPopFunc` if you define it like `final FutureOr<bool> Function()? canPopFunc;`.  

Now you may not be able to do that, just thinking out loud.

**Comment by michael-woolsey** [comment]
Dialogs appear to be closing on the Android back button without navigating when I use the `QDialog` widget which is awesome!

Testing ongoing...

**Comment by michael-woolsey** [comment]
@SchabanBo - I think I found an issue, I have a route that is similar to this:
```
    QRoute(
      path: '/some/route',
      middleware: [
        QMiddlewareBuilder(canPopFunc: () => quitConfirmationDialog2()),
      ],
      builder: () => const SomeScreen(),
    ),
```
And on a screen within that route I have a button that does this:
```
OutlinedButton(
  onPressed: () => QR.navigator.replaceAll('/landing'),
  child: Text(translate('instructionsPager.quit')),
)
```

When I click the quit button, the dialog does indeed popup as expected, however when I click `no` to not quit it the app seems to go into some form of loop causing the dialog to keep appearing no matter how many times I click no.

Clicking `yes` navigates me as expected, but I cannot seem to click `no` and simply have the dialog close.

Hopefully this makes sense.

**Comment by SchabanBo** [comment]
@michael-woolsey Can you please show me  (quitConfirmationDialog2)

**Comment by michael-woolsey** [comment]
```
Future<bool> quitConfirmationDialog2() async {
  final result = await QR.show<bool?>(
    QDialog(
      widget: (pop) {
        return StatefulBuilder(
          builder: (BuildContext context, void Function(void Function()) setState) {
            final translate = AppLocalizations.of(context)!.translate;

            return AlertDialog(
              title: Text(translate('quitConfirmationDialog.title')),
              contentPadding: const EdgeInsets.fromLTRB(24.0, 20.0, 24.0, 12.0),
              content: Container(
                width: 300,
                constraints: const BoxConstraints(maxHeight: 300),
                child: SingleChildScrollView(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      Padding(
                        padding: const EdgeInsets.symmetric(vertical: 8.0),
                        child: Text(translate('quitConfirmationDialog.message')),
                      ),
                    ],
                  ),
                ),
              ),
              actions: [
                TextButton(
                  onPressed: () => pop(false),
                  child: Text(translate('quitConfirmationDialog.noButton').toUpperCase()),
                ),
                Container(
                  width: 8.0,
                ),
                TextButton(
                  onPressed: () => pop(true),
                  child: Text(translate('quitConfirmationDialog.yesButton').toUpperCase()),
                ),
              ],
            );
          },
        );
      },
    ),
  );

  print('result: $result');
  return result ?? false;
}```

**Comment by michael-woolsey** [comment]
The only reason I'm using a `StatefulBuilder` is because I need a `BuildContext` for my translations and that seemed like the easiest way to get one.

**Comment by SchabanBo** [comment]
@michael-woolsey 
can you please test this issue in the dev branches and tell me if it is solved?
```yaml
     qlevar_router:
       git:
         url: https://github.com/SchabanBo/qlevar_router.git
         ref: develop
```

**Comment by michael-woolsey** [comment]
@SchabanBo 

There are two scenarios I'm testing to ensure they both work and unfortunately one is still broken.  So I will attempt to tell you what I'm seeing.

1. When I press a `Quit` button, I have coded it so that it navigates using `QR.navigator.replaceAll('/landing')`.
- [X] I can press `yes` and it goes to the landing page, as expected
- [X] I can press `no` and it stays where I am, as expected
- [X] I press the `back button` it stays where I am, as expected

2. On the same screen as no. 1 I have the ability to press the back button which should take me to the previous screen.
- [X] I can press `yes` and it goes to the previous screen, as expected
- [ ] When I press `no` however, it 'closes' the app down and it's no longer visible
- [X] I can press the `back button` and it stays where I am, as expected

So it seems there is still a problem when clicking `no` when I used the `back button` to attempt to navigate.

Both scenario 1 and scenario 2 both use the same dialog as listed above `quitConfirmationDialog2`.

If you need any more information, please let me know.

**Comment by SchabanBo** [comment]
@michael-woolsey
Can you please test the dev branch again?
I had to rewrite the Pop logic a little bit. But it should work as expected now.

**Comment by michael-woolsey** [comment]
I did a `flutter clean` and a `pub get` and it has the same issue, I'm doing the following steps:

1. Navigate to `screen1` (Uses `QR.to('/screen1')`)
2. Press a button to navigate to `screen2` (Uses `QR.to('/screen2')`)
```
 QRoute(
  path: '/screen2',
  middleware: [
    QMiddlewareBuilder(canPopFunc: () => quitConfirmationDialog2()),
  ],
  builder: () => const Screen2(),
),
```
3. When I press the `android back button` to go back to `screen1` I get the `quitConfirmationDialog2()` dialog as expected.
4. Next press the `no` button `onPressed: () => pop(false),` expecting that I should stay on the screen.
5. The app disappears off the screen, like it got minimized.  It's still running in the background but it should have simply closed the dialog and that's it.

This is the `quitConfirmationDialog2()` that I am using:
```
Future<bool> quitConfirmationDialog2() async {
  final result = await QR.show<bool?>(
    QDialog(
      widget: (pop) {
        return StatefulBuilder(
          builder: (BuildContext context, void Function(void Function()) setState) {
            final translate = AppLocalizations.of(context)!.translate;

            return AlertDialog(
              title: Text(translate('quitConfirmationDialog.title')),
              contentPadding: const EdgeInsets.fromLTRB(24.0, 20.0, 24.0, 12.0),
              content: Container(
                width: 300,
                constraints: const BoxConstraints(maxHeight: 300),
                child: SingleChildScrollView(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      Padding(
                        padding: const EdgeInsets.symmetric(vertical: 8.0),
                        child: Text(translate('quitConfirmationDialog.message')),
                      ),
                    ],
                  ),
                ),
              ),
              actions: [
                TextButton(
                  onPressed: () => pop(false),
                  child: Text(translate('quitConfirmationDialog.noButton').toUpperCase()),
                ),
                Container(
                  width: 8.0,
                ),
                TextButton(
                  onPressed: () => pop(true),
                  child: Text(translate('quitConfirmationDialog.yesButton').toUpperCase()),
                ),
              ],
            );
          },
        );
      },
    ),
  );

  print('result: $result');
  print('actual result: ${result ?? false}');
  return result ?? false;
}
```

I even tried removing the `StatefulBuilder` from the `quitConfirmationDialog2()` dialog to see if that's the issue somehow, but the problem persists.

**Comment by michael-woolsey** [comment]
@SchabanBo 
I have to pass this issue off to a co-worker of mine to continue working on it as I've have to work on something else with my company.  

I've asked him to post a message here so that he's part of the ticket as well.

Thank you for your effort and hopefully we can find the issue soon!

**Comment by lowkwp** [comment]
Hi @SchabanBo, I'm the co-worker of @michael-woolsey.

I just tested the develop branch and it worked as expected. 
It seems that `flutter clean` and `pub get` didn't reflect your changes (regarding `PopResult`).

Thank you for your effort.

**Comment by michael-woolsey** [comment]
Oh awesome find @lowkwp!  Thanks for catching that!

**Comment by SchabanBo** [comment]
@michael-woolsey @lowkwp
Great, glad to help. I will release these changes as 1.5.1

**Comment by lowkwp** [comment]
Hi @SchabanBo 

Sorry to call you here again.
I think I found an issue with closing dialog with android back button.
Generally, the back button works well as expected, but it closes the app when the dialog is opened from the first route.
In other words, when there is just one route in the route tree, `QR.back` seems to return `NotPoped` instead of `DialogClosed`.

This happens in the first screen of our app and in the screen accessed with `replaceAll`.

Could you verify this?

Thank you.

**Comment by SchabanBo** [comment]
@lowkwp Nice catch.
This should be fixed now with 1.5.1. 
I will wait for your feedback.

**Comment by lowkwp** [comment]
@SchabanBo 
Now it works as expected.
Thank you for the effort!
